### PR TITLE
Feature/2111 toggle header options

### LINKED ISSUE #2111: Show/Hide Table Controls
## Summary

Provide the ability to selective hide table controls like:
1. Search box
2. Filters
3. Data download
4. Pagination controls

## Motivation

There are use cases where one would want to disable search/filtering/downloading of data or even hide the page next previous controls as one might be doing more complex pagination by themselves.

### PR DESCRIPTION
**Pull request by wmdev0808**
## Description
- Add the header options section for the visibility on the property pane
- Render conditionally the header part based on the setting

Fixes #2111 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feature/2111-toggle-header-options 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 51.68 **(0.02)** | 32.85 **(0.01)** | 29.59 **(0.02)** | 52.15 **(0.01)**
 :red_circle: | app/client/src/components/designSystems/appsmith/TableComponent/TableStyledWrappers.tsx | 35.37 **(0)** | 15.38 **(-0.37)** | 0 **(0)** | 35.37 **(0)**
 :green_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 64.2 **(0.22)** | 52.31 **(-0.01)** | 54.17 **(0)** | 63.58 **(0.24)**
 :green_circle: | app/client/src/utils/migrations/TableWidget.ts | 78.7 **(3.17)** | 63.46 **(2.59)** | 78.57 **(3.57)** | 81 **(2.84)**
 :red_circle: | app/client/src/widgets/TableWidget/index.tsx | 9.31 **(-0.15)** | 0 **(0)** | 10.64 **(0)** | 9.97 **(-0.24)**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
The feature works now. But there is an issue with old tables. There are too many extra rows. This doesn't happen for the new table widgets dropped on the canvas. @riodeuno Don't know if some kinda migration is required here.

@vivekverma2312 can we test this Deploy preview thoroughly.

**Comment by vivekverma2312** [comment]
sure @somangshu I will start testing this now.

**Comment by vivekverma2312** [comment]
@somangshu Issues observed so far:
1. By default Header is showing disabled for old tables
2. Extra rows are coming in the table for the below cases:
  -Existing table in the preview
  -If I create a table on release env and check on preview
3. First row height changes when all the header property is in disabled state

Screenshots for Issue 3:
Screenshot 2021-05-20 at 2 17 47 PM
Screenshot 2021-05-20 at 2 16 31 PM

**Comment by somangshu** [comment]
> 1. By default Header is showing disabled for old tables

@riodeuno will need some help here. How do we migrate the old table for this case?

File: app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx [hunk 1]
```diff
@@ -161,65 +166,87 @@ export function Table(props: TableProps) {
...
-        ref={tableHeaderWrapperRef}
-        serverSidePaginationEnabled={props.serverSidePaginationEnabled}
-        tableSizes={tableSizes}
-        width={props.width}
-      >
-        <Scrollbars
-          renderThumbHorizontal={ScrollbarHorizontalThumb}
-          renderThumbVertical={ScrollbarVerticalThumb}
-          style={{ width: props.width, height: 38 }}
+      {(props.isVisibleSearch ||
```
Comment (SatishGandham) [inline app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx:186]: Can we use isHeaderVisible here?

**Review by SatishGandham — CHANGES_REQUESTED** [review]
Looks good except for the small change requested.

File: app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx (hunk 1 above)
Comment (wmdev0808) [reply app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx]: I've made a new PR for your comment

**Review by somangshu — DISMISSED** [review]
Approve to run test

**Comment by somangshu** [comment]
@wmdev0808 some related tests are failing here! Can you check

**Comment by wmdev0808** [comment]
> @wmdev0808 some related tests are failing here! Can you check

Hi @somangshu, I've reviewed, but I don't think it is directly related to the current PR.
I can see a few failing cypress tests, but I can't get the point

**Comment by somangshu** [comment]
@wmdev0808 some test fails here because the tests cannot find the table header. [Check this out](https://dashboard.cypress.io/projects/eyxvp8/runs/6555/overview). 

@vicky-primathon can you take a look?

**Comment by vicky-primathon** [comment]
Sure @somangshu looking into it

**Review by SatishGandham — DISMISSED** [review]

**Comment by vicky-primathon** [comment]
@wmdev0808  @somangshu  we are hiding table header section by default, this is causing table test cases to fail. Can we enable the table header toggles by default

**Comment by somangshu** [comment]
@vicky-primathon this needs migration I believe. For an old table widget, the header options are hidden by default.

Please help @wmdev0808 with this!

**Comment by vicky-primathon** [comment]
Okay @somangshu

**Review by vicky-primathon — APPROVED** [review]
Looks good to me

**Comment by vivekverma2312** [comment]
Tested Deploy Preview and working fine. Good to merge into release.

**Comment by somangshu** [comment]
@yatinappsmith @NandanAnantharamu tests failing here! `ForkApplication_spec.js` is not related. The last release merge was two days ago.

**Review by somangshu — APPROVED** [review]

**Pull request merged**